src/utils/AssayGeneratorMetabo.cpp

```diff
@@ -0,0 +1,523 @@
+    // Parsing parameters
+    //-------------------------------------------------------------
+
+    String in = getStringOption_("in");
+    String id = getStringOption_("in_id");
+    String out = getStringOption_("out");
+    String method = getStringOption_("method");
+    double precursor_mz_tol = getDoubleOption_("precursor_mz_tolerance");
+    double precursor_rt_tol = getDoubleOption_("precursor_rt_tolerance");
+    double cosine_sim_threshold = getDoubleOption_("cosine_similarity_threshold");
+    bool method_consensus_spectrum = method == "consensus_spectrum" ? true : false;
+
+    unsigned int hull_size_filter = getIntOption_("filter_by_convex_hulls");
+    double transition_threshold = getDoubleOption_("transition_threshold");
+    bool use_known_unknowns = getFlag_("use_known_unknowns");
+
+    // load mzML
+    MzMLFile mzml;
+    PeakMap spectra;
+    mzml.load(in, spectra);
+
+    // load featurexml
+    FeatureXMLFile fxml;
+    FeatureMap feature_map;
+    fxml.load(id, feature_map);
+
+    // check if correct featureXML is given and set use_known_unkowns parameter if no id information is available
+    const std::vector<DataProcessing>& processing = feature_map.getDataProcessing();
+    for (auto it = processing.begin(); it != processing.end(); ++it)
+    {
+      if (it->getSoftware().getName() == "FeatureFinderMetabo")
+      {
+        // check if convex hulls parameter was used in the FeatureFinderMetabo
+        if (it->getMetaValue("parameter: algorithm:ffm:report_convex_hulls") != "true")
+        {
+          throw Exception::InvalidParameter(__FILE__,
+                                            __LINE__,
+                                            OPENMS_PRETTY_FUNCTION,
+                                            "Please provide a valid feature XML file with reported convex hulls.");
+        }
+        // if id information is missing set use_known_unknowns to true
+        if (feature_map.getProteinIdentifications().empty())
+        {
+          use_known_unknowns = true;
+        }
+      }
+    }
+
+    // annotate precursor mz and intensity
+    annotatePrecursorIntensity(spectra, 0.2, 0.2);
+
+    // filter feature by convexhull size
+    auto map_it = remove_if(feature_map.begin(), feature_map.end(),
+                                                 [&hull_size_filter](const Feature& f) -> bool
+                                                 {
+                                                   return f.getConvexHulls().size() < hull_size_filter;
+                                                 });
+    feature_map.erase(map_it, feature_map.end());
+
+    KDTreeFeatureMaps fp_map_kd;
+    vector<FeatureMap> v_fp;
+    v_fp.push_back(feature_map);
+    fp_map_kd.addMaps(v_fp);
+
+    // read FeatureMap in KDTree for feature-precursor assignment
+    map<const BaseFeature*, vector<size_t> > map_feature_to_precursors = extractMetaInformation(spectra, fp_map_kd, precursor_mz_tol, precursor_rt_tol);
```

Comment:
this is a bit confusing as you are actually mapping features to triggered MS2 spectra (while the feature represents the precursor). I suggest to rename this to `feature_ms2_spectra_map` or something similar